libp2p/protocols/pubsub/pubsubpeer.nim

```diff
@@ -338,10 +345,18 @@ proc sendEncoded*(p: PubSubPeer, msg: seq[byte], isHighPriority: bool): Future[v
         libp2p_gossipsub_priority_queue_size.inc(labelValues = [$p.peerId])
     f
   else:
-    let f = p.rpcmessagequeue.nonPriorityQueue.addLast(msg)
-    when defined(pubsubpeer_queue_metrics):
-      libp2p_gossipsub_non_priority_queue_size.inc(labelValues = [$p.peerId])
-    f
+    if len(p.rpcmessagequeue.nonPriorityQueue) == p.maxNumElementInNonPriorityQueue:
+      if not p.disconnected:
```

Comment:
if this flag is set, it seems prudent to check in `connectOnce` that we don't re-attempt connection, ie `DisconnectionRequested` should lead to the send conn never being opened again

Reply:
thanks, I saw that peers kept being disconnected multiple times and that was probably the reason. But I had to add the check in `connectImpl`.

Reply:
well, the non-prio queue should also be cleared (ie all sends in it should be failed) at this point since it will no longer be sent - and if there are any new sends after the disconnection, they should fail

Reply:
also, it _maybe_ makes sense to balance the streamOpened event with a streamclosed in this case, such events typically benefit from being neutralised this way

Reply:
Canceling the future will cause this problem again https://github.com/vacp2p/nim-libp2p/pull/1075, won't it?

Reply:
> maybe makes sense to balance the streamOpened event with a streamclosed in this case

Could you please elaborate more on what you mean by that?

Reply:
> Canceling the futures will cause this problem again #1075, won't it?

Not sure if you meant canceling the futures. `stopSendNonPriorityTask` clears the queue and it's called in `unsubscribePeer` which should be called when the peer is disconnected.

Reply:
> Not sure if you meant canceling the futures. 

I meant failing them, not cancelling, but now I realise we always return a completed future instead of storing pending futures in the queue, so there's nothing to fail:

https://github.com/vacp2p/nim-libp2p/blob/unstable/libp2p/protocols/pubsub/pubsubpeer.nim#L341

`addLast` always returns a completed future when the async queue is unbounded (as it is in this case).

Reply:
> Should I also emit a Streamclosed?

this is an idea for a future improvement - it's maybe not needed right now, however

Reply:
There's basically a lifecycle here:

`created -> [stream opened -> stream closed]* -> disconnected` where the middle two states repeat, so when designing an event api around it, these are the relevant events.